iceoryx_posh/source/popo/active_call_set.cpp

```diff
@@ -0,0 +1,223 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/popo/active_call_set.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/event_listener.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/event_notifier.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/cxx/helplets.hpp"
+
+namespace iox
+{
+namespace popo
+{
+ActiveCallSet::ActiveCallSet() noexcept
+    : ActiveCallSet(runtime::PoshRuntime::getInstance().getMiddlewareEventVariable())
+{
+}
+
+ActiveCallSet::ActiveCallSet(EventVariableData* eventVariable) noexcept
+    : m_eventVariable(eventVariable)
+    , m_eventListener(*eventVariable)
+{
+    m_thread = std::thread(&ActiveCallSet::threadLoop, this);
+}
+
+ActiveCallSet::~ActiveCallSet()
+{
+    m_wasDtorCalled.store(true, std::memory_order_relaxed);
+    m_eventListener.destroy();
+
+    m_thread.join();
+    m_eventVariable->m_toBeDestroyed.store(true, std::memory_order_relaxed);
+}
+
+cxx::expected<uint64_t, ActiveCallSetError>
+ActiveCallSet::addEvent(void* const origin,
+                        const uint64_t eventType,
+                        const uint64_t eventTypeHash,
+                        CallbackRef_t<void> callback,
+                        TranslationCallbackRef_t translationCallback,
+                        const cxx::MethodCallback<void, uint64_t> invalidationCallback) noexcept
+{
+    std::lock_guard<std::mutex> lock(m_addEventMutex);
+
+    for (uint64_t i = 0U; i < MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET; ++i)
+    {
+        if (m_events[i]->isEqualTo(origin, eventType, eventTypeHash))
+        {
+            return cxx::error<ActiveCallSetError>(ActiveCallSetError::EVENT_ALREADY_ATTACHED);
+        }
+    }
+
+    uint32_t index = 0U;
+    if (!m_indexManager.pop(index))
+    {
+        return cxx::error<ActiveCallSetError>(ActiveCallSetError::ACTIVE_CALL_SET_FULL);
+    }
+
+    m_events[index]->init(index, origin, eventType, eventTypeHash, callback, translationCallback, invalidationCallback);
+    return cxx::success<uint64_t>(index);
+}
+
+void ActiveCallSet::removeEvent(void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) noexcept
+{
+    for (uint32_t index = 0U; index < MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET; ++index)
+    {
+        if (m_events[index]->resetIfEqualTo(origin, eventType, eventTypeHash))
+        {
+            m_indexManager.push(index);
+            break;
+        }
```

Comment:
is it possible that two threads call `removeEvent` with the same parameters?

Reply:
Yes it is allowed or lets say not explicitly forbidden. But `m_events` is covered by a `smart_lock`. For fun let's say you do this, calling this at with the same parameters at the same time.
In line 80:
`if (m_events[index]->resetIfEqualTo(origin, eventType, eventTypeHash))` the `resetIfEqualTo` is only called by one thread at a time since the `smart_lock` is securing this call till the end of the statement. The thread which wins resets the index successfully and is now able to push the index back into the indexmanager.

The next one who calls this right after the first thread would then return false and would never return the index then.

Reply:
This is a typical race condition in the algorithm of the user. The active call set never reaches a undefined state or some internal inconsistency.
I added a detailed doxygen comment describing this particular scenario but here I would not restrict the usage of the active call set in any way or refine the algorithm. In the end it is the same problem like:
```cpp
if ( !myConcurrentMultiPusherQueue.empty() ) {
  // anything can happen between the empty call and pop since we are working 
  // concurrently on this queue. Therefore, the user has to take care of this problem in his
  // algorithm.
  auto element = myConcurrentMultiPusherQueue.pop();
}
```